### PR TITLE
Switch to go modules

### PR DESCRIPTION
**Pull request by saschagrunert**
Hey, this PR switch our dependency handling mechanism to go modules.

Things which are done here:
- Update packages which case the `s/Sirupsen` dependency name clash
- Vendor in all external tooling dependencies which usually would be retrieved via `go get`
- Remove `vndr`
- Update `make vendor` to use go modules
- Added the `export GO111MODULE=off` to the `Makefile`, which allows to build cri-o offline per default

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -175,13 +177,9 @@ ${GINKGO}: ${BUILD_BIN_PATH}
...
-vendor: .install.vndr
-	$(GOPATH)/bin/vndr \
-		-whitelist github.com/onsi/ginkgo \
-		-whitelist github.com/golang/mock \
-		-whitelist github.com/cpuguy83/go-md2man \
-		-whitelist github.com/russross/blackfriday \
-		${PKG}
+vendor:
+	$(GO) mod tidy
+	$(GO) mod vendor
```
Comment (vrothberg) [inline Makefile]: `$(GO) mod verify` to be ultra sure nobody tricked us :)

Comment (saschagrunert) [reply Makefile]: Alright, I added it :+1:

**Comment by saschagrunert** [comment]
> Thanks for kicking this off! I planned to do this for CRI-O in the near future as we have to go top down our dependency tree and migrate the projects step-by-step. But I am happy somebody else volunteers :)

This is really kinda fun. I found out which packages caused the name clash with `s/Sirupsen` and updated them. We also have now the possibility to vendor external needed tools via the hack in `tools/tools.go`. So there should be no need for `go get` invocations during CI any more.

**Comment by saschagrunert** [comment]
@chavafg Do you think we could update the go version on the kata runner, this should fix this issue:

```
go: verifying k8s.io/client-go@v0.0.0-20190228174230-b40b2a5939e4: checksum mismatch
	downloaded: h1:DOmAZxy5pQ64Zz3qmDEmf5sFEQvYswldnGtrwiDx5hE=
	go.sum:     h1:aE8wOCKuoRs2aU0OP/Rz8SXiAB0FTTku3VtGhhrkSmc=
```

**Comment by chavafg** [comment]
Hi @saschagrunert, we currently use 1.11.1 defined [here](https://github.com/kata-containers/runtime/blob/master/versions.yaml#L272).
I think we can update it. To which version? 1.12.3, which is the latest? 
/cc @sboeuf

**Comment by saschagrunert** [comment]
> I think we can update it. To which version? 1.12.3, which is the latest?

Yep that would be great. If you want than I can file the PR too.

**Comment by chavafg** [comment]
I see you already opened a PR, thanks.

**Comment by mrunalp** [comment]
Is this ready?

**Comment by saschagrunert** [comment]
> Is this ready?

From my side yes. :)

**Comment by mrunalp** [comment]
@cblecker Not sure why the dco check is still pending here.

**Comment by mrunalp** [comment]
@saschagrunert @vrothberg If we have to vendor cri-o packages into a project that doesn't use go modules, will that cause an issue?

**Comment by saschagrunert** [comment]
> @saschagrunert @vrothberg If we have to vendor cri-o packages into a project that doesn't use go modules, will that cause an issue?

It should be unproblematic. If there is no go.mod file within a dependency then go uses the source code discovery to obtain needed paths. For example we vendor `github.com/BurntSushi/toml` which does not contain a go.mod/sum file. I also added `GO111MODULE=off` to the Makefile to avoid the internet lookup/retrieval during build. So our repo is as self contained as before.

One thing worth to note is that I updated all dependencies (like docker) which relied on the old Sirupsen/logrus package. We’re also now on Kubernetes 1.14.0 but this should be fine too, right?

**Comment by mrunalp** [comment]
@saschagrunert Thanks for the explanation :) Yeah k8s 1.14 is fine. We really want to switch to 1.15-dev version but that can land after.

**Comment by mrunalp** [comment]
@saschagrunert One more thing we need to fix. I am seeing this on a local build:
```
[mrunalp@localhost cri-o]$ sudo make install
[sudo] password for mrunalp: 
go build -i -ldflags '-s -w -X main.gitCommit="af269e22426064bba72fd009d1f96142767d7a1e" -X main.buildInfo=1555108575' -tags "     seccomp selinux  containers_image_ostree_stub" -o bin/crio github.com/cri-o/cri-o/cmd/crio
make -C conmon
make[1]: Entering directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/conmon'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/conmon'
make -C pause
make[1]: Entering directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/pause'
make[1]: '../bin/pause' is up to date.
make[1]: Leaving directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/pause'
install -Z -D -m 755 bin/crio /usr/local/bin/crio
install -Z -D -m 755 bin/conmon /usr/local/libexec/crio/conmon
install -Z -D -m 755 bin/pause /usr/local/libexec/crio/pause
mkdir -p /build/bin
go build -o /build/bin/go-md2man ./vendor/github.com/cpuguy83/go-md2man
vendor/github.com/cpuguy83/go-md2man/md2man.go:9:2: cannot find package "github.com/cpuguy83/go-md2man/md2man" in any of:
        /usr/lib/golang/src/github.com/cpuguy83/go-md2man/md2man (from $GOROOT)
        /home/mrunalp/go/src/github.com/cri-o/cri-o/_output/src/github.com/cpuguy83/go-md2man/md2man (from $GOPATH)
make: *** [Makefile:176: /build/bin/go-md2man] Error 1

```

**Comment by saschagrunert** [comment]
> @saschagrunert One more thing we need to fix. I am seeing this on a local build:
> 
> ```
> [mrunalp@localhost cri-o]$ sudo make install
> [sudo] password for mrunalp: 
> go build -i -ldflags '-s -w -X main.gitCommit="af269e22426064bba72fd009d1f96142767d7a1e" -X main.buildInfo=1555108575' -tags "     seccomp selinux  containers_image_ostree_stub" -o bin/crio github.com/cri-o/cri-o/cmd/crio
> make -C conmon
> make[1]: Entering directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/conmon'
> make[1]: Nothing to be done for 'all'.
> make[1]: Leaving directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/conmon'
> make -C pause
> make[1]: Entering directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/pause'
> make[1]: '../bin/pause' is up to date.
> make[1]: Leaving directory '/home/mrunalp/go/src/github.com/cri-o/cri-o/pause'
> install -Z -D -m 755 bin/crio /usr/local/bin/crio
> install -Z -D -m 755 bin/conmon /usr/local/libexec/crio/conmon
> install -Z -D -m 755 bin/pause /usr/local/libexec/crio/pause
> mkdir -p /build/bin
> go build -o /build/bin/go-md2man ./vendor/github.com/cpuguy83/go-md2man
> vendor/github.com/cpuguy83/go-md2man/md2man.go:9:2: cannot find package "github.com/cpuguy83/go-md2man/md2man" in any of:
>         /usr/lib/golang/src/github.com/cpuguy83/go-md2man/md2man (from $GOROOT)
>         /home/mrunalp/go/src/github.com/cri-o/cri-o/_output/src/github.com/cpuguy83/go-md2man/md2man (from $GOPATH)
> make: *** [Makefile:176: /build/bin/go-md2man] Error 1
> ```

Yes, I see. This issue originates from the go-md2-man vendoring change previously. I have a fix in mind and will follow up in a separate PR.

**Comment by cblecker** [comment]
@saschagrunert Is this a build tool that you were vendoring?

**Comment by saschagrunert** [comment]
> @saschagrunert Is this a build tool that you were vendoring?

Yes, it is the build of the github.com/cpuguy83/go-md2man tool, which does not work any more when running under sudo because of invalid path expansions.

**Comment by cblecker** [comment]
Have you tried explicitly prefixing `GOMODULE111=off` to your build command? It might work to shut modules off, just for that command.

**Comment by saschagrunert** [comment]
> Have you tried explicitly prefixing `GOMODULE111=off` to your build command? It might work to shut modules off, just for that command.

We have the GOMODULE111=off for mainly all `Makefile` targets except to `make vendor` to disable remote lookups druing the build and force go to use the vendor dir. The issue is mainly related to an invalid build path when running one of the make targets via `sudo`. Nevertheless, thanks for looking into the issue and supporting here, that's nice. :)

**Comment by cblecker** [comment]
The other thing we’ve done sometimes over in some Kubernetes scripts is refer to the package by its full gopath, instead of relative directory.. e.g. `github.com/cri-o/cri-o/vendor/github.com/cpuguy83/go-md2man`. Just one other lead that might be worth trying, off the top of my head.

**Comment by saschagrunert** [comment]
> The other thing we’ve done sometimes over in some Kubernetes scripts is refer to the package by its full gopath, instead of relative directory.. e.g. `github.com/cri-o/cri-o/vendor/github.com/cpuguy83/go-md2man`. Just one other lead that might be worth trying, off the top of my head.

Indeed that works too! Now I have two fixes for one problem. :) I will prefer your mentioned approach.

**Pull request merged**